# Analyzing Used Ebay Car Sales Data

We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The version of the dataset we are working with is a sample of 50,000 data points that was prepared by [Dataquest](https://www.dataquest.io/) including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which which year the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which which year the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos=pd.read_csv("autos.csv", encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50


Our dataset contains 20 columns, most of which are stored as strings. There are a few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with.

## Formatting Columns

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

A quick review of the column names shows that some column names are in camelcase as opposed to Python's preferred snakecase. Since there are no spaces or other common delimiters in any of the column names, we won't be able use the split() command to replace the spaces or delimiters with underscores. Instead, we will snake-case the column names and replace them. We will also reword some column names to more closely decribe the type of data they contain.

In [7]:
autos.columns=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Prelimnary Data Exploration and Cleaning

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In the first phase of our data exploration and cleaning process, let's try to find columns that have mostly one value and drop them. A quick review of the output tells us the columns seller, offertype and nr_of_pictures are the candidates that fit the bill.


In [10]:
autos.drop(['seller','offer_type','nr_of_pictures'],axis=1,inplace=True)

In [11]:
autos.dtypes

date_crawled          object
name                  object
price                 object
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer              object
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
postal_code            int64
last_seen             object
dtype: object

There are two columns, price and auto, which are numeric values with extra characters being stored as text. We'll clean and convert these.

In [12]:
autos['price']=autos['price'].str.replace("$","").str.replace(",","").astype(float)

In [13]:
autos['odometer']=autos['odometer'].str.replace("km","").str.replace(",","").astype(int)

In [14]:
autos.rename({'price':'price_usd','odometer':"odometer_km"},axis=1,inplace=True)

In [15]:
autos.dtypes

date_crawled           object
name                   object
price_usd             float64
abtest                 object
vehicle_type           object
registration_year       int64
gearbox                object
power_ps                int64
model                  object
odometer_km             int32
registration_month      int64
fuel_type              object
brand                  object
unrepaired_damage      object
ad_created             object
postal_code             int64
last_seen              object
dtype: object

Now let's explore the price and odometer columns

In [16]:
autos[["odometer_km","price_usd"]].describe()

odometer_km     price_usd
count   50000.000000  5.000000e+04
mean   125732.700000  9.840044e+03
std     40042.211706  4.811044e+05
min      5000.000000  0.000000e+00
25%    125000.000000  1.100000e+03
50%    150000.000000  2.950000e+03
75%    150000.000000  7.200000e+03
max    150000.000000  1.000000e+08

In [17]:
autos['price_usd'].value_counts().shape[0]

2357

In [18]:
autos['price_usd'].value_counts().sort_index(ascending=True).head(20)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price_usd, dtype: int64

The prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.

There are 1,421 cars listed with $0 price - given that this is only 2% of the of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [19]:
autos['price_usd'].value_counts().sort_index(ascending=False).head(15)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price_usd, dtype: int64

There are a number of listings with prices below 30 including about 1,500 at 0. There are also a small number of listings with very high values, including 14 at around or over 

Given that eBay is an auction site, there could legitimately be items where the opening bid is 1 . We will keep the 1 dollar items, but remove anything above $350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [20]:
autos=autos[autos['price_usd'].between(1,350000)]

Let's look at the odometer column

In [21]:
autos['odometer_km'].unique().shape

(13,)

In [22]:
autos['odometer_km'].value_counts().sort_index(ascending=True).head(15)

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.None of the values seem out of the ordinary so we will leave this column as is.

## Exploring Dates

There are a number of columns with date information:

- date_crawled
- registration_month
- registration_year
- ad_created
- last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.

In [23]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [24]:
autos['date_crawled'].str[:10].value_counts(dropna=False,normalize=True).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [25]:
autos['date_crawled'].str[:10].value_counts(dropna=False,normalize=True).sort_values()

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [26]:
autos['ad_created'].str[:10].value_counts(dropna=False,normalize=True).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [27]:
autos['last_seen'].str[:10].value_counts(dropna=False,normalize=True).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future

## Dealing with Incorrect Registration Year Data

Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [28]:
(~autos['registration_year'].between(1900,2016)).sum()/autos.shape[0]

0.038793369710697

Given that this is less than 4% of our data, we will remove these rows.

In [29]:
autos=autos.loc[autos['registration_year'].between(1900,2016)]
autos['registration_year'].value_counts(normalize=True).sort_values(ascending=False).head(20)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

## Exploring Price by Brand

In [30]:
((autos['brand'].value_counts(normalize=True))*100).sort_values(ascending=False)

volkswagen        21.126368
bmw               11.004477
opel              10.758124
mercedes_benz      9.646323
audi               8.656627
ford               6.989996
renault            4.714980
peugeot            2.984083
fiat               2.564212
seat               1.827296
skoda              1.640925
nissan             1.527388
mazda              1.518819
smart              1.415994
citroen            1.400998
toyota             1.270324
hyundai            1.002549
sonstige_autos     0.981127
volvo              0.914719
mini               0.876159
mitsubishi         0.822604
honda              0.784045
kia                0.706926
alfa_romeo         0.664082
porsche            0.612669
suzuki             0.593389
chevrolet          0.569825
chrysler           0.351321
dacia              0.263490
daihatsu           0.250637
jeep               0.227073
subaru             0.214220
land_rover         0.209936
saab               0.164949
jaguar             0.156381
daewoo             0

In [31]:
brand_distribution=((autos['brand'].value_counts(normalize=True))*100).sort_values(ascending=False)


German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [32]:
price_by_brand={}
for car in brand_distribution[brand_distribution>5].index:
    mean_price=autos.loc[autos['brand']==car,'price_usd'].mean()
    price_by_brand[car]=int(mean_price)


price_by_brand


{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

Of the top 6 brands, there is a distinct price gap:

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

In [33]:
mileage_by_brand={}
for car in brand_distribution[brand_distribution>5].index:
    #mean_price=autos.loc[autos['brand']==car,'price_usd'].mean()
    mean_mileage=autos.loc[autos['brand']==car,'odometer_km'].mean()
    
    mileage_by_brand[car]=int(mean_mileage)
mileage_by_brand

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [34]:
price_by_brand_series=pd.Series(price_by_brand).sort_values(ascending=False)
mileage_by_brand_series=pd.Series(mileage_by_brand)


In [35]:
mean_price_mileage_by_brand_df=pd.DataFrame(price_by_brand_series,columns=['mean_price'])

In [36]:
mean_price_mileage_by_brand_df['mean_mileage']=mileage_by_brand_series

In [37]:
mean_price_mileage_by_brand_df

mean_price  mean_mileage
audi                 9336        129157
mercedes_benz        8628        130788
bmw                  8332        132572
volkswagen           5402        128707
ford                 3749        124266
opel                 2975        129310

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands.

## Ninja Level Cleaning

This is stricly for cleaning-ninjas. Faint hearted peeps, you have been forewarned! What follows is the product of my obsessive need to go above and beyond in assignemnts and solve all the additional questions that have been asked. So without wasting another minute, brooms, kick - GO! #HP4Life

###  Identifying Categorical Data

The first activity requires identifying columns with categorical values and translating them to English. The reason for the transalation is because the data is from Ebay Germany and so they are in German. Here is a quick snapshot of the data to reaquaint you with what is in it:

In [38]:
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price_usd   abtest vehicle_type  registration_year    gearbox  power_ps  \
0     5000.0  control          bus               2004    manuell       158   
1     8500.0  control    limousine               1997  automatik       286   
2     8990.0     test    limousine               2009    manuell       102   
3     4350.0  control   kleinwagen               2007  automatik        71   
4     1350.0     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

So here are categorical columns that need to be translated: gearbox,fuel_type, and unrepaired_damage

In [92]:
gearbox_gr_en={"manuell":"manual","automatik":"automatic"}

In [96]:
#autos['gearbox']=autos['gearbox'].map(gearbox_gr_en) #uncomment when re-running code

In [122]:
autos['gearbox'].value_counts(dropna=False)

manual       34715
automatic     9856
NaN           2110
Name: gearbox, dtype: int64

In [119]:
fuel_type_gr_en={"benzin":"petrol","diesel":"diesel","lpg":"lpg","cng":"cng","hybrid":"hybrid","elektro":"electric","andere":"other"}
#autos['fuel_type']=autos['fuel_type'].map(fuel_type_gr_en) #uncomment when re-running code
autos['fuel_type'].value_counts(dropna=False)

petrol      28540
diesel      14032
NaN          3318
lpg           649
cng            71
hybrid         37
electric       19
other          15
Name: fuel_type, dtype: int64

In [123]:
autos['unrepaired_damage'].value_counts(dropna=False)

nein    33834
NaN      8307
ja       4540
Name: unrepaired_damage, dtype: int64

In [124]:
unrep_damage_gr_en={"nein":"no","ja":"yes"}
autos['unrepaired_damage']=autos['unrepaired_damage'].map(unrep_damage_gr_en) #uncomment when re-running code
autos['unrepaired_damage'].value_counts(dropna=False)

no     33834
NaN     8307
yes     4540
Name: unrepaired_damage, dtype: int64

## Standardizng Dates

Let's reformat dates and convert them to integers.

In [125]:
autos['ad_created'].head()

0    2016-03-26 00:00:00
1    2016-04-04 00:00:00
2    2016-03-26 00:00:00
3    2016-03-12 00:00:00
4    2016-04-01 00:00:00
Name: ad_created, dtype: object

In [135]:
autos['ad_created'].str[:10].str.replace("-","").astype(int)

0        20160326
1        20160404
2        20160326
3        20160312
4        20160401
           ...   
49995    20160327
49996    20160328
49997    20160402
49998    20160308
49999    20160313
Name: ad_created, Length: 46681, dtype: int32

In [136]:
autos['date_crawled'].str[:10].str.replace("-","").astype(int)

0        20160326
1        20160404
2        20160326
3        20160312
4        20160401
           ...   
49995    20160327
49996    20160328
49997    20160402
49998    20160308
49999    20160314
Name: date_crawled, Length: 46681, dtype: int32

In [137]:
autos['last_seen'].str[:10].str.replace("-","").astype(int)

0        20160406
1        20160406
2        20160406
3        20160315
4        20160401
           ...   
49995    20160401
49996    20160402
49997    20160404
49998    20160405
49999    20160406
Name: last_seen, Length: 46681, dtype: int32

## Most Common Brands/Models

In [138]:
autos['brand-model']=autos['brand']+'-'+autos['model']

In [143]:
autos['brand-model'].value_counts().sort_values(ascending=False).head(10)

volkswagen-golf           3707
bmw-3er                   2615
volkswagen-polo           1609
opel-corsa                1592
volkswagen-passat         1349
opel-astra                1348
audi-a4                   1231
mercedes_benz-c_klasse    1136
bmw-5er                   1132
mercedes_benz-e_klasse     958
Name: brand-model, dtype: int64

VW's Golf is the most popular car in the dataset. Out of the top 5, 3 are VW's models. This is line with what we have seen before. VW is clearly the most popular choice of car buyers owing to the price and performance balance they offer. 

## Mileage vs Price Variation

In [157]:
sorted(autos['odometer_km'].unique())

[5000,
 10000,
 20000,
 30000,
 40000,
 50000,
 60000,
 70000,
 80000,
 90000,
 100000,
 125000,
 150000]

In [175]:
conditions=[(autos['odometer_km'].between(5000,10000)),
 (autos['odometer_km'].between(10000,30000)),
 (autos['odometer_km'].between(3000,60000)),
 (autos['odometer_km'].between(60000,90000)),
 (autos['odometer_km']>90000)]

tiers=["5000-10000","10000-30000","30000-60000","60000-90000","> 90000"]
autos['odometer_tiers']=np.select(conditions,tiers)

In [180]:
autos['odometer_tiers'].value_counts(dropna=False)

> 90000        37000
60000-90000     4235
30000-60000     2918
10000-30000     1502
5000-10000      1026
Name: odometer_tiers, dtype: int64

In [181]:
odometer_tiers_index=autos['odometer_tiers'].value_counts(dropna=False).index

In [183]:
mileage_price={}
for i in odometer_tiers_index:
    mean_price_by_mileage=autos.loc[autos['odometer_tiers']==i,'price_usd'].mean()
    mileage_price[i]=mean_price_by_mileage
mileage_price
    

{'> 90000': 4331.8022432432435,
 '60000-90000': 9563.2186540732,
 '30000-60000': 13721.36017820425,
 '10000-30000': 17517.63382157124,
 '5000-10000': 11616.441520467837}

We can see a direct inverse relation between price and mileage. Higher the mileage, lower the price of the car. 

## Repaired vs Unreparied Price Variation

In [184]:
autos[]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price_usd   abtest vehicle_type  registration_year    gearbox  \
0         5000.0  control          bus               2004     manual   
1         8500.0  control    limousine               1997  automatic   
2         8990.0     test    limousine               2009     manual   
3         4350.0  control   kleinwagen               2007  automatic   
4         1350.0     test        kombi               2003     manual   
...          ...      ...          ...                ...        ...   
49995    24900.0  control    limousine               2011  automatic   
49996     1980.0  control       cabrio               1996     manual   
49997    13200.0     test       cabrio               2014  automatic   
49998    22900.0  control        kombi               2013     manual   
49999     1250.0  control    limousine               1996     manual   

       power_ps   model  odometer_km  registration_month fuel_type  \
0           158  andere       150000                   3       lpg   
1           286     7er       150000                   6    petrol   
2           102    golf        70000                   7    petrol   
3            71  fortwo        70000                   6    petrol   
4             0   focus       150000                   7    petrol   
...         ...     ...          ...                 ...       ...   
49995       239      q5       100000                   1    diesel   
49996        75   astra       150000                   5    petrol   
49997        69     500         5000                  11    petrol   
49998       150      a3        40000                  11    diesel   
49999       101  vectra       150000                   1    petrol   

            brand unrepaired_damage           ad_created  postal_code  \
0         peugeot                no  2016-03-26 00:00:00        79588   
1             bmw                no  2016-04-04 00:00:00        71034   
2      volkswagen                no  2016-03-26 00:00:00        35394   
3           smart                no  2016-03-12 00:00:00        33729   
4            ford                no  2016-04-01 00:00:00        39218   
...           ...               ...                  ...          ...   
49995        audi                no  2016-03-27 00:00:00        82131   
49996        opel                no  2016-03-28 00:00:00        44807   
49997        fiat                no  2016-04-02 00:00:00        73430   
49998        audi                no  2016-03-08 00:00:00        35683   
49999        opel                no  2016-03-13 00:00:00        45897   

                 last_seen      brand-model odometer_tiers  
0      2016-04-06 06:45:54   peugeot-andere        > 90000  
1      2016-04-06 14:45:08          bmw-7er        > 90000  
2      2016-04-06 20:15:37  volkswagen-golf    60000-90000  
3      2016-03-15 03:16:28     smart-fortwo    60000-90000  
4      2016-04-01 14:38:50       ford-focus        > 90000  
...                    ...              ...            ...  
49995  2016-04-01 13:47:40     

In [216]:
mean_price_unrepaired_cars=autos.loc[autos['unrepaired_damage']=='yes','price_usd'].mean()
mean_price_repaired_cars=autos.loc[autos['unrepaired_damage']=='no','price_usd'].mean()
pct_difference=((mean_price_repaired_cars-mean_price_unrepaired_cars)/mean_price_unrepaired_cars)*100
times_difference=round(mean_price_repaired_cars/mean_price_unrepaired_cars)

In [233]:
print("Average price of repaired cars is : {:.2f}".format(mean_price_repaired_cars))
print("Average price of unrepaired cars is : {:.2f}".format(mean_price_unrepaired_cars))
print("Un-repaired cars are cheaper than repaired cars by {0:.2f}% (i.e by {1} times)".format(pct_difference,times_difference))

Average price of repaired cars is : 7164.03
Average price of unrepaired cars is : 2241.15
Un-repaired cars are cheaper than repaired cars by 219.66% (i.e by 3 times)


## Conclusion

- German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand with the VW Golf being the most popular model
- The brand of a car does not have much of an effect on the milelage of the cars.
- We can see a direct inverse relation between price and mileage. Higher the mileage, lower the price of the car.
- Un-repaired cars are 3 times cheaper than their repaired counter parts
